File: js/Game.js
var question = require('./question.js');

function Game(playerName, locations, inventory, turns){
	this.playerName = playerName;
	this.location = 0;
	this.locations = locations();
	this.locationBuilder = locations;
	this.inventory = inventory;
	this.turns = turns;
	this.currentDay = 0;
}

Game.prototype.day = function(){
	this.currentDay++;
	if(this.currentDay < this.turns){
		this.updateMarkets();
		this.startDay();
	}else{
		this.gameOver();
	}
};

Game.prototype.startDay = function(){
	console.log('\n\n-----------------------------------');
	console.log('Oh hai, ' + this.playerName + '!');
	console.log('day ' + this.currentDay);
	console.log('days left: ' + (this.turns - this.currentDay));
	this.mainLoop();
};

Game.prototype.mainLoop = function(){
	this.status();
	this.outputTradeOptions();
};

Game.prototype.gameOver = function(){
	console.log('\nGame Over!!');
	console.log('You earned £' + this.cash + '\n');
};

Game.prototype.status = function(){
	var currentLocation = this.getLocation();
	console.log('\nInventory:');
	console.log(this.inventory.display());
	console.log(currentLocation.display());

};

Game.prototype.outputTradeOptions = function(){

	var purchaseOptions = this.getLocationItems(0);
	purchaseOptions = purchaseOptions.concat(this.getInventoryItems(purchaseOptions.length));
	purchaseOptions.push(this.getTravelAction('m'));
	purchaseOptions.push(this.getQuitAction('q'));

	question.multi('Would you like to trade? ', purchaseOptions);

};

Game.prototype.buyQuantity = function(item){
	question.ask('how many? ', function(answer){
		try{			
			var price = item.price * answer;
			
			if(this.inventory.cash - price < 0){
				console.log('not enough money');
			}else{
				// TODO : see below
				//this.inventory.purchase(item.name, parseInt(answer, 10), price);

				this.inventory.cash -= item.price * answer;	
				this.inventory.add(item.name, parseInt(answer, 10));
				this.getLocation().sell(item.name, answer);
			}
			
		}catch(e){
			console.log('there are not that many units available!\n');
		}

		this.mainLoop();
	}.bind(this));
};

Game.prototype.sellQuantity = function(item){
	question.ask('how many? ', function(answer){

		try{
			this.inventory.remove(item.name, answer);
			var receipt = this.getLocation().buy(item.name, parseInt(answer, 10));

			this.inventory.cash += receipt;
		}catch(e){
			console.log('there are not that many units available!\n');
		}

		this.mainLoop();
	}.bind(this));
};

Game.prototype.outputMovementOptions = function(){

	var otherLocations = this.locations.filter(function(item){
		return item !== this.getLocation();
	}.bind(this));

	var options = otherLocations.map(function(location, index){
		return {
			code: index,
			description: location.name,
			callback: this.updateLocation.bind(this, location)
		};
	}.bind(this));

	options.push(this.getQuitAction('q'));

	question.multi('Where would you like to go? ', options);
};

Game.prototype.updateMarkets = function() {
	this.locations = this.locationBuilder();
};

Game.prototype.updateLocation = function(location){
	this.location = this.locations.indexOf(location);
	this.day();
};

Game.prototype.getLocation = function(){
	return this.locations[this.location];
};

Game.prototype.getLocationItems = function(startCode){
	var availableItems = this.getLocation().getItems()
	var options = availableItems.map(function(item){
		return {
			code: startCode++,
			description: 'buy ' + item.name,
			callback:this.buyQuantity.bind(this, item)
		};
	}.bind(this));

	return options;
};

Game.prototype.getInventoryItems = function(startCode){
	var options = this.inventory.items.map(function(item){
			return {
				code: startCode++,
				description:'sell ' + item.name,
				callback: this.sellQuantity.bind(this, item)
			};
	}.bind(this));

	return options;
};

Game.prototype.getTravelAction = function(code){
	return {
		code: code,
		description: 'move location',
		callback: this.outputMovementOptions.bind(this)
	};
};

Game.prototype.getQuitAction = function(code){
	return {
		code: code,
		description: 'quit',
		callback: function(){
			console.log('goodbye!');
			process.exit();
		}
	};
};


module.exports = Game;

File: python/Game.py
# This Python file uses the following encoding: utf-8
import os
from multiChoice import ask
from Location import Location
from Inventory import Inventory
from random import randrange


class Game:
	
	currentDay = 0
	currentLocation = 0
	def __init__(self, name, locations, cash, products, days):
		self.maxDays = days
		self.locations = getLocations(locations, products)
		self.name = name;
		self.inventory = Inventory(cash)
		self.startingCash = cash

	def start(self):
		self.currentDay += 1
		location = self.getLocation()
		print 'Oh hai {0}!'.format(self.name)
		print '{0} days left!'.format(self.maxDays - self.currentDay)
		print 'Location: {0}'.format(location.name)
		if(self.currentDay >= self.maxDays):
			return self.gameOver()
		self.mainLoop()

	def mainLoop(self):		
		self.inventory.display() 
		self.getLocation().display()
		self.getTradeOptions()

	def gameOver(self):
		print 'Game Over!'
		print 'You finished with £{0}'.format(self.inventory.cash)
		print 'Minus the £{0} you have to repay your mum, that\'s £{1}!'.format(self.startingCash, self.inventory.cash - self.startingCash)

	def getLocation(self):
		return self.locations[self.currentLocation]

	def getTradeOptions(self):
		code = 0
		options = []
		location = self.getLocation()
		for product in location.market:
			option = {'code': str(code), 'description':product['name'], 'callback':self.getPurchaseClosure(product)}
			options.append(option)
			code += 1
		options.append({'code':'m', 'description':'check out a different area', 'callback':self.travel})
		options.append({'code':'q', 'description':'quit', 'callback':self.exit})
		ask('Would you like to trade?', options)
	def getPurchaseClosure(self, product):
		def cb():
			qty = raw_input('how many?\n')
			self.purchase(product, int(qty))
		return cb
	def exit(self):
		print 'goodbye!'
		os._exit(0)	
	def travel(self):
		code = 0
		options = []
		for location in self.locations:
			option = {'code': str(code), 'description':location.name, 'callback':self.getMovementClosure(location)}
			options.append(option)
			code += 1
		options.append({'code':'q', 'description':'quit', 'callback':self.exit})
		ask('Where would you like to go?', options)
	def getMovementClosure(self, location):
		def cb():
			self.moveTo(location)
		return cb
	def moveTo(self, location):
		self.currentLocation = self.locations.index(location)
		self.start()
	def purchase(self, product, qty):
		price = product['price'] * qty
		if(self.inventory.cash >= price and product['quantity'] >= qty):
			self.inventory.cash -= price
			self.inventory.add(product['name'], qty)
			product['quantity'] -= qty
		elif(self.inventory.cash < price):
			print 'insufficient funds'
			print 'you only have £{0}, but transaction costs £{1}'.format(self.inventory.cash, price)
		else:
			print 'not that many available'
			print 'you want {0}, but there are only {1} units'.format(qty, product['quantity'])
		self.mainLoop()
		


""" location builder functions """

def getLocations(locationNames, products):	
	locationList = []
	for location in locationNames:
		productList = setupProducts(products)
		locationList.append(Location(location, productList))
	return locationList

def setupProducts(productDefs):
	products = []
	for product in productDefs:
		products.append({
				'name': product['name'],
				'price': getPrice(product),
				'quantity':getQuantity(),
				'min':product['min'],
				'max':product['max']
			})
	return products


def getPrice(product):
	return randrange(product['min'], product['max'], 1)

def getQuantity():
	return 10